### ISSUE TITLE
[Bug] F.grid_sample backward pass on CUDA returns accumulated gradients due to uninitialized grad_input memory

### ISSUE THREAD

**Issue by xyt66665000** [pytorch/pytorch#185495]
### 🐛 Describe the bug

When running `torch.nn.functional.grid_sample` on CUDA, independent backward passes on newly created input tensors appear to produce gradients that accumulate across runs.

In the reproducer below, each iteration creates a fresh leaf tensor `x` with `requires_grad=True`, runs `F.grid_sample`, and calls `backward()` once. However, `x.grad[0, 0, 0, 0]` increases as `1.0, 2.0, 3.0, ...` across independent iterations.

Since the input tensor is newly created in every iteration, its `.grad` field should not carry over from previous iterations. The observed monotonic accumulation suggests that stale data may be present in the internal `grad_input` buffer used by the CUDA backward implementation.

### Minimum Reproducible Example

```python
import torch
import torch.nn.functional as F
import gc

def reproduce_uninitialized_memory():
    device = "cuda"
    dtype = torch.float32

    print(f"PyTorch Version: {torch.__version__}")
    
    grid = torch.tensor([[[[-1.0, -1.0], [ 1.0, -1.0]],
                          [[-1.0,  1.0], [ 1.0,  1.0]]]], device=device, dtype=dtype)
    
    corner_gradients = []

    print("\nStarting independent forward and backward passes...")
    
    for i in range(1, 6):
        x = torch.zeros(1, 1, 4, 4, device=device, dtype=dtype, requires_grad=True)
        
        y = F.grid_sample(x, grid, align_corners=True)
        
        y.sum().backward()
        
        val = x.grad[0, 0, 0, 0].item()
        corner_gradients.append(val)
        print(f"Run {i}: Top-left gradient = {val}")
        
        del x
        del y
        gc.collect()

    if len(set(corner_gradients)) > 1 and corner_gradients[-1] > corner_gradients[0]:
        print("\nBug Reproduced: Gradients are accumulating across independent runs.")
    else:
        print("\nBug Not Reproduced: Gradients are consistent.")

if __name__ == "__main__":
    reproduce_uninitialized_memory()
```

### Expected behavior
Each iteration creates a fresh input tensor and performs one backward pass. The top-left input gradient should be consistently `1.0` in every run.

### Actual behavior
The top-left input gradient increases across independent runs:

1.0, 2.0, 3.0, 4.0, 5.0

This looks like stale gradient data is being reused or accumulated into, rather than starting from a zero-initialized gradient buffer.

Output from the nightly build:
```text
(torch-nightly) xyt19@Oasis:/tmp$ python bug.py
PyTorch Version: 2.13.0.dev20260521+cu130

Starting independent forward and backward passes...
Run 1: Top-left gradient = 1.0
Run 2: Top-left gradient = 2.0
Run 3: Top-left gradient = 3.0
Run 4: Top-left gradient = 4.0
Run 5: Top-left gradient = 5.0

Bug Reproduced: Gradients are accumulating across independent runs.
```

### Versions

PyTorch version: 2.13.0.dev20260521+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.0.140
Nvidia driver version: 596.49
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_engines_tensor_ir.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.21.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.21.1
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A
ersions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0.dev20260521+cu130
[pip3] torchaudio==2.11.0.dev20260525+cu130
[pip3] torchvision==0.28.0.dev20260525+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] numpy 2.2.6 pypi_0 pypi
[conda] nvidia-cublas 13.1.1.3 pypi_0 pypi
[conda] nvidia-cuda-cupti 13.0.85 pypi_0 pypi
[conda] nvidia-cuda-nvrtc 13.0.88 pypi_0 pypi
[conda] nvidia-cuda-runtime 13.0.96 pypi_0 pypi
[conda] nvidia-cudnn-cu13 9.20.0.48 pypi_0 pypi
[conda] nvidia-cufft 12.0.0.61 pypi_0 pypi
[conda] nvidia-curand 10.4.0.35 pypi_0 pypi
[conda] nvidia-cusolver 12.0.4.66 pypi_0 pypi
[conda] nvidia-cusparse 12.6.3.3 pypi_0 pypi
[conda] nvidia-cusparselt-cu13 0.8.1 pypi_0 pypi
[conda] nvidia-nccl-cu13 2.29.7 pypi_0 pypi
[conda] nvidia-nvjitlink 13.0.88 pypi_0 pypi
[conda] nvidia-nvtx 13.0.85 pypi_0 pypi
[conda] torch 2.13.0.dev20260521+cu130 pypi_0 pypi
[conda] torchaudio 2.11.0.dev20260525+cu130 pypi_0 pypi
[conda] torchvision 0.28.0.dev20260525+cu130 pypi_0 pypi
[conda] triton 3.7.0+git88b227e2 pypi_0 pypi

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

**Comment by xyt66665000** [comment]
I also reproduced this with `torch.autograd.grad`, which avoids writing into `x.grad` and rules out user-side `.grad` accumulation.

Output:
```text
PyTorch Version: 2.13.0.dev20260521+cu130
CUDA Version: 13.0

Run 1: Top-left gradient = 1.0
Run 2: Top-left gradient = 2.0
Run 3: Top-left gradient = 3.0
Run 4: Top-left gradient = 4.0
Run 5: Top-left gradient = 5.0

Observed gradients: [1.0, 2.0, 3.0, 4.0, 5.0]
````

Since `autograd.grad` returns a fresh gradient tensor instead of accumulating into `x.grad`, this suggests the accumulated value is produced inside the CUDA `grid_sample` backward computation itself.

**Comment by xyt66665000** [comment]
I've confirmed that this exact same uninitialized memory bug also affects dtype=torch.float64. This confirms the issue is agnostic to the floating-point precision and sits purely at the tensor allocation/initialization level.

**Comment by eqy** [comment]
Can you share what gpu you used? I cannot reproduce this on B200.

**Comment by xyt66665000** [comment]
Hi @eqy  @soulitzer ,

I finally figured out why this was so hard to reproduce on your end! The root cause is actually tied to the exact same environment conflict I discovered in my other issue: #185512.

In #185512, I found that having a global cuDNN 9.21.1 installed via apt alongside PyTorch's pip-installed cuDNN 9.20.0 wheel caused F.conv2d to crash with a CUDNN_STATUS_SUBLIBRARY_VERSION_MISMATCH because the dynamic loader was pulling libcudnn_engines_tensor_ir.so.9 from the system path.

It turns out that this exact same library mismatch causes silent memory corruption/gradient accumulation in F.grid_sample.

Here is the exact proof. When I restore the system-level cuDNN, the bug appears:
```text
xyt19@Oasis:/tmp$ sudo apt install \
  libcudnn9-cuda-12=9.21.1.3-1 \
  libcudnn9-dev-cuda-12=9.21.1.3-1 \
  libcudnn9-headers-cuda-12=9.21.1.3-1
[sudo] password for xyt19:
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  libcudnn9-cuda-12 libcudnn9-dev-cuda-12 libcudnn9-headers-cuda-12
0 upgraded, 3 newly installed, 0 to remove and 77 not upgraded.
Need to get 482 MB of archives.
After this operation, 1107 MB of additional disk space will be used.
Get:1 https://developer.download.nvidia.cn/compute/cuda/repos/ubuntu2404/x86_64  libcudnn9-cuda-12 9.21.1.3-1 [482 MB]
Get:2 https://developer.download.nvidia.cn/compute/cuda/repos/ubuntu2404/x86_64  libcudnn9-headers-cuda-12 9.21.1.3-1 [35.2 kB]
Get:3 https://developer.download.nvidia.cn/compute/cuda/repos/ubuntu2404/x86_64  libcudnn9-dev-cuda-12 9.21.1.3-1 [12.9 kB]
Fetched 482 MB in 2min 24s (3351 kB/s)
Selecting previously unselected package libcudnn9-cuda-12.
(Reading database ... 122477 files and directories currently installed.)
Preparing to unpack .../libcudnn9-cuda-12_9.21.1.3-1_amd64.deb ...
Unpacking libcudnn9-cuda-12 (9.21.1.3-1) ...
Selecting previously unselected package libcudnn9-headers-cuda-12.
Preparing to unpack .../libcudnn9-headers-cuda-12_9.21.1.3-1_amd64.deb ...
Unpacking libcudnn9-headers-cuda-12 (9.21.1.3-1) ...
Selecting previously unselected package libcudnn9-dev-cuda-12.
Preparing to unpack .../libcudnn9-dev-cuda-12_9.21.1.3-1_amd64.deb ...
+ exit 0
Unpacking libcudnn9-dev-cuda-12 (9.21.1.3-1) ...
Setting up libcudnn9-headers-cuda-12 (9.21.1.3-1) ...
Setting up libcudnn9-cuda-12 (9.21.1.3-1) ...
Setting up libcudnn9-dev-cuda-12 (9.21.1.3-1) ...
xyt19@Oasis:/tmp$ conda activate torch-nightly
(torch-nightly) xyt19@Oasis:/tmp$ python bug.py
PyTorch Version: 2.13.0.dev20260521+cu130

Starting independent forward and backward passes...
Run 1: Top-left gradient = 1.0
Run 2: Top-left gradient = 2.0
Run 3: Top-left gradient = 3.0
Run 4: Top-left gradient = 4.0
Run 5: Top-left gradient = 5.0

Bug Reproduced: Gradients are accumulating across independent runs.
```

But as soon as I remove the system-level cuDNN (so PyTorch only uses its bundled pip cuDNN libraries), the gradients are completely correct:
```text
xyt19@Oasis:/tmp$ sudo apt remove libcudnn9-cuda-12 libcudnn9-dev-cuda-12 libcudnn9-headers-cuda-12
[sudo] password for xyt19:
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages will be REMOVED:
  libcudnn9-cuda-12 libcudnn9-dev-cuda-12 libcudnn9-headers-cuda-12
0 upgraded, 0 newly installed, 3 to remove and 77 not upgraded.
After this operation, 1107 MB disk space will be freed.
Do you want to continue? [Y/n] y
(Reading database ... 122527 files and directories currently installed.)
Removing libcudnn9-dev-cuda-12 (9.21.1.3-1) ...
Removing libcudnn9-cuda-12 (9.21.1.3-1) ...
Removing libcudnn9-headers-cuda-12 (9.21.1.3-1) ...
xyt19@Oasis:/tmp$ conda activate torch-nightly
(torch-nightly) xyt19@Oasis:/tmp$ python bug.py
PyTorch Version: 2.13.0.dev20260521+cu130

Starting independent forward and backward passes...
Run 1: Top-left gradient = 1.0
Run 2: Top-left gradient = 1.0
Run 3: Top-left gradient = 1.0
Run 4: Top-left gradient = 1.0
Run 5: Top-left gradient = 1.0

Bug Not Reproduced: Gradients are consistent.
```
While F.conv2d correctly throws a loud mismatch error (#185512), F.grid_sample silently produces mathematically incorrect gradients (stale memory accumulation) under the exact same conflicting cuDNN environment.

Since the root cause is a local environment dependency conflict, I can close this issue if you consider this strictly a user-side environment problem. However, I wanted to report this because a silent correctness/uninitialized memory bug caused by library fallbacks feels quite dangerous. Let me know how you'd like to proceed!

**Comment by jbschlosser** [comment]
Closing this issue due to environment-related issues. If the problem is found to be reproducible outside of these issues, it can be reopened.
